storage/interface.go

```diff
@@ -255,6 +255,92 @@ func (f QueryableFunc) Querier(mint, maxt int64) (Querier, error) {
 	return f(mint, maxt)
 }
 
+// AppendSampleOptions provides optional auxiliary data and configuration for AppenderV2.AppendSample
+type AppendSampleOptions struct {
+	// Ref (optional) represents the storage SeriesRef for the given series, if known. This helps to avoid extra lookup work.
+	Ref SeriesRef
+
+	// MetricFamilyName (optional) provides metric family name for the appended sample's
+	// series. If the client of the AppenderV2 has this information
+	// (e.g. from scrape) it's recommended to pass it to the appender.
+	//
+	// Provided strin bytes are unsafe to reuse, it only lives for the duration of the AppendSample call.
+	//
+	// Some implementations use this avoid slow and prone to error metric family detection for:
+	// * Metadata per metric family storages (e.g. Prometheus metadata WAL/API/RW1)
+	// * Strictly complex types storages (e.g. OpenTelemetry Collector).
+	//
+	// NOTE(krajorama): Example purpose is highlighted in OTLP ingestion: OTLP calculates the
+	// metric family name for all metrics and uses it for generating summary,
+	// histogram series by adding the magic suffixes. The metric family name is
+	// passed down to the appender in case the storage needs it for metadata updates.
+	// Known user of this is Mimir that implements /api/v1/metadata and uses
+	// Remote-Write 1.0 for this. Might be removed later if no longer
+	// needed by any downstream project.
+	// NOTE(bwplotka): Long term, once Prometheus uses complex types on storage level
+	// the MetricFamilyName can be removed as MetricFamilyName will equal to __name__ always.
+	MetricFamilyName string
+
+	// Metadata (optional) attached to the appended sample.
+	// Metadata strings are safe for reuse.
+	Metadata metadata.Metadata
+
+	// Exemplars (optional) attached to the appended sample.
+	// Exemplar slice is unsafe for reuse.
+	Exemplars []exemplar.Exemplar
+
+	// DiscardOutOfOrder TBD
+	DiscardOutOfOrder bool
+}
+
+// NoOpts represents empty AppendSampleOptions options.
+// This is for readability only.
+var NoOpts = AppendSampleOptions{}
```

Comment:
Maybe the better approach is to have some alias on `var AppendSampleOptions = ASOpts`